docs/Workflow.md

```diff
@@ -51,102 +95,180 @@ git remote set-url --push upstream no_push
 
-Get your local main up to date:
+At the end of the repository setup commands above:
+- `upstream` will refer to the original ONNX-MLIR repository.
+- `origin` will refer to your own fork of ONNX-MLIR.
 
+Never commit anything to your fork's `main` branch, the only way you should update `main` is from `upstream`. The procedure to update your fork's `main` branch is listed in Step 4.
+
+## Step 4: Keeping your repository up to date
+
+To keep your ONNX-MLIR fork's `main` up to date perform the following steps:
+
+1. Fetch the latest versions of your fork (`origin`) and the `upstream` repositories:
+```sh
+git fetch --all
+```
+
+2. Update the `main` branch on your fork:
 ```sh
-cd $working_dir/onnx-mlir
-git fetch upstream
 git checkout main
-git rebase upstream/main
+git pull
+git merge upstream/main
+git push origin main
 ```
 
-Branch from main:
+Provided you have never committed anything to your fork's `main` branch directly, all the updates to your fork's `main` should be fast forwards.
 
+3. The `main` branch of your fork should now be identical to the `main` branch of `upstream`. To check you can do:
 ```sh
-git checkout -b myfeature
+git diff upstream/main
 ```
+and the command will return immediately signaling that no differences exist between `upstream/main` and `origin/main`
 
-### Step 4: Develop
+4. (Optional) Assuming your `main` is up to date, to update any branches you are currently working on to use the latest ONNX-MLIR, you need to do the following:
 
-#### Edit the code
+```sh
+git checkout my-branch
+git merge origin/main
+```
 
-You can now edit the code on the `myfeature` branch.
+If no conflicts are signaled and the merge is complete do:
 
-#### Run cmake & make
+```sh
+git push origin my-branch
+```
 
-Follow the directions to build ONNX-MLIR in this section of the main 
-[README](../README.md#onnx-mlir-this-project)
+However, if conflicts appear, the merge will be interrupted until the conflicts are resolved. A list of files will be marked as containing conflicts. To identify those files do:
 
+```sh
+git status -uno
+```
 
-#### Run Test
+The files in red are the files containing conflicts. Go to all the files which contain a conflict and resolve the conflicts.
+When the conflicts are resolved do a `git add` on each conflicted file:
 
 ```sh
-# Run unit test to make sure all test passed.
-make check-onnx-lit
-make check-onnx-backend
-make check-onnx-backend-dynamic
-make check-onnx-backend-constant
-make test
+git add path/to/file1
+git add path/to/file2
+...
 ```
 
-### Step 5: Commit
-
-Commit your changes, always using the `-s` flag in order to sign your commits.
+When all conflicted files have been added do:
+```sh
+git commit
+```
 
+Followed by a git push:
 ```sh
-git commit -s
+git push origin my-branch
 ```
 
-Likely you'll go back and edit/build/test some more than `commit --amend -s`
-in a few cycles.
+Your branch is now up to date with the latest ONNX-MLIR.
+
+## Step 5: Create a branch for your changes
```

Comment:
we should not speak about `your branch` until after this step where we explain how to build one...